server/aesdsocket.c

```diff
@@ -168,227 +494,148 @@ int main(int argc , char *argv[])
+        int new_fd = accept(sd, (struct sockaddr *)&client_addr, &addr_size);
 
-        // Write the data packet to the file
-        int buffer_len = ((datapacket + 1) + (BUFFER_SIZE * realloc_int));
-        int write_len = write(fd , write_buffer , buffer_len);
         // Error check
-        if(write_len == -1){
-            syslog(LOG_ERR , "write\n");
-            return -1;
+        if (ERROR == new_fd) {
+            if (errno == EWOULDBLOCK) {
+                
+                continue;
+            }
+            syslog(LOG_ERR , "accept\n");   
+            continue;
         }
-        // printf("Write len:%d\n", write_len);
 
-        file_size += write_len;
-
-        // printf("Size: %d\n", file_size);
-
-        // Reset the reallocation tracking flag and the final size tracking flag
-        realloc_int = 0, final_size = 1;
-
-        // Allocate memory for the send buffer
-        char *send_buffer = (char *)malloc(sizeof(char) * file_size);
+        // Malloc new node
+        node_t *new_node = (node_t *)malloc(sizeof(node_t));
         // Error check
-        if(send_buffer == NULL){
+        if (new_node == NULL){
             syslog(LOG_ERR , "malloc\n");
-            return -1;
+            return ERROR;
         }
 
-        // Go to the beginning of the file using lseek
-        off_t seek_status = lseek(fd , 0 , SEEK_SET);
-        // Error check
-        if(seek_status == -1){
-            syslog(LOG_ERR , "lseek\n");
-            return -1;
-        }
+        // Initialize and add values of the attributes to the new node
+        new_node->thread_data.is_complete = 0;
+        new_node->thread_data.conn_fd = new_fd;
+        new_node->thread_data.client_addr = (struct sockaddr_in *)&client_addr;
 
-        // Read the contents of the file
-        ret_status = read(fd , send_buffer , file_size);
+        // Spawn a new thread to handle the new connection and write the data from that connection
+        ret_status = pthread_create(&(new_node->thread_data.tid), NULL, socketThreadfunc, &(new_node->thread_data));
         // Error check
-        if(ret_status == -1){
-            syslog(LOG_ERR , "read\n");
-            return -1;
-        }
+        if(ret_status != SUCCESS){
+            syslog(LOG_ERR , "pthread_create\n");
+            printf("Error in pthread_create with err number: %d", errno);
+        } 
 
-        // Send the buffer to the client
-        ret_status = send(new_fd , send_buffer , file_size , 0);
-        if(ret_status == -1){
-            syslog(LOG_ERR , "send\n");
-            return -1;
+        // Add the new node to the head of the linked list
+        addFirst(&head, new_node); 
+    }
+    // Iterating logic to remove from linked list
+    current = head;
+    previous = head;
+
+    while(current != NULL) {
+        // When the node is done executing and is the head
+        if ((current->thread_data.is_complete == true) && (current == head)) {
```

Comment:
@KhyatiSatta see discussion in class about this.  I'd suggest using this `is_complete` logic in the loop which starts threads.  Without this you will not be removing old threads until the program exits.  You also need cleanup here which `pthread_join`'s threads regardless of completion state, waiting for completion as needed.